src/persist-txn/src/txn_read.rs

```diff
@@ -365,26 +388,51 @@ impl<T: Timestamp + Lattice + TotalOrder + StepForward + Codec64, C: TxnsCodec>
 
     async fn update<F: Fn(&T) -> bool>(&mut self, done: F) {
         while !done(&self.progress_exclusive) {
-            let events = self.txns_subscribe.fetch_next().await;
+            let events = self.txns_subscribe.next().await;
             for event in events {
-                let mut updates = match event {
+                let parts = match event {
                     ListenEvent::Progress(frontier) => {
                         self.progress_exclusive = frontier
                             .into_option()
                             .expect("nothing should close the txns shard");
                         continue;
                     }
-                    ListenEvent::Updates(updates) => updates,
+                    ListenEvent::Updates(parts) => parts,
                 };
+                let mut updates = Vec::new();
+                for part in parts {
+                    let should_fetch_part = self.should_fetch_part(&part);
+                    debug!(
+                        "should_fetch_part={} for {:?} {:?}",
+                        should_fetch_part,
+                        self.only_data_id,
+                        part.stats()
+                    );
+                    if !should_fetch_part {
+                        self.txns_subscribe.return_leased_part(part);
+                        continue;
+                    }
+                    let part_updates = self.txns_subscribe.fetch_batch_part(part).await;
+                    // To keep memory usage down, filter out unrelated data
+                    // shards even before we buffer them for the timestamp sort.
```

Comment:
This comment should probably move upward, close to the filtering it describes!

Reply:
Heh, funny enough this is a second filtering pass. I'll add a comment for the pushdown one above, too